### DISCUSSION TITLE
ViperIDE: `.mpy` disassembler

### DISCUSSION THREAD

**Discussion by vshymanskyy** [micropython/micropython#15576] [General]
Working on `.mpy` disassembler in [ViperIDE](https://viper-ide.org).
Cutting-edge or cringe? Tell me! 😅
Would it be useful?

![image](https://github.com/user-attachments/assets/9c4623ad-c4f4-4b94-bf40-d8dce1630e21)

**Comment by GitHubsSilverBullet** [comment]
> Cutting-edge or cringe? Tell me!

IMNSHO: Definitely cringe.
Those who create a mpy file have the original. So there's no point except – let's just called it what it is – for those who want to steal. Just .02

**Reply by ned-pcs** [reply]
There are already tools that allow disassembly of `.mpy` files.
One obvious example is `mpy-tool` in the MicroPython `/tools` directory.

**Reply by vshymanskyy** [reply]
Which is exactly what is running here, it's `mpy-tool`. No plans to do anything beyond that

**Reply by GitHubsSilverBullet** [reply]
@vshymanskyy You asked for an opinion. You've got one now.
All you have to provide in response are snide remarks.
Maybe you'd like to think about why a GUI tool is like an open invitation while a command line tool is not.
Plonk!

**Reply by vshymanskyy** [reply]
Sorry, an open invitation to do what exactly? To explore your binary packages? To educate yourself?

**Comment by vshymanskyy** [comment]
I'm wondering if this could be useful for advanced developers or at least for educational purposes?
Maybe it could drive some interest in MPY internals

**Reply by robert-hh** [reply]
I would consider it as over feature-ism, especially for an IDE aiming at a clean and slick design. People interested in MPY internals can use mpy-tool themselves, and they should be able to do so.

**Reply by vshymanskyy** [reply]
I'm not quite there yet, but looks like I can have this feature "almost for free".
Clean design has nothing to do with it. Currently, `mpy` files are displayed as hex dumps, it could be a disassembled form as well.

**Reply by vshymanskyy** [reply]
Well, I can have a separate checkbox in the Settings tab for this. Default to off for the peace of mind of @GitHubsSilverBullet

**Reply by GitHubsSilverBullet** [reply]
Seriously? Do you have some kind of a problem? A personal vendetta maybe?
Man, live with it that others may have different opinions than you.
And if you can't, don't ask.

**Reply by robert-hh** [reply]
@GitHubsSilverBullet @vshymanskyy Gents, please pace your words, stay polite.

**Reply by vshymanskyy** [reply]
I'm just curious since when dissecting a binary using standard tools file became a crime 🤪

**Reply by rkompass** [reply]
I find it useful for educational purposes.
Advanced developers and those who intend to steal will use `mpy-tool`, of course. So I can't find a problem here.
My impression is that other features would be more desirable first, but who am I to make you prescriptions in which order you implement new things?
More features are welcome here as long as the IDE doesn't get cluttered.

Is it possible/acceptable for you @vshymanskyy to delete the post with
> Looks like @GitHubsSilverBullet is really worried ..

?
I have the impression it is offending. Referring to something quite unrelated. Without necessity.

**Reply by vshymanskyy** [reply]
Ok, removed it. Sorry if i crossed a blurry line

**Reply by vshymanskyy** [reply]
@rkompass this is exactly the output of `mpy-tool` with some additional syntax highlight on top of it.
The reason i did it, it's a part of broader concept - self-hosting MicroPython tools under MicroPython VM (running in browser in this instance). I think this unlocks reusability of tools and supports the ecosystem

**Reply by rkompass** [reply]
I also have the experience that luxury like that coloring can make things easier and in the end help you to be more productive. Thanks.

**Reply by vshymanskyy** [reply]
> My impression is that other features would be more desirable first

@rkompass what do you have on your mind?
(except being able to sync the device FS with a local folder or GitHub)

**Reply by rkompass** [reply]
A few more file operations:
- Renaming of files/folders, perhaps as option of the delete confirming pop-up dialog.
- Not exactly syncing, but almost: Now that creating a folder with '+' works, how about creating/importing a folder with the same '+', using a button at the pop-up dialog for the filename which allows to select a folder from the local file system?

A way to display graphics:
Often MP is used to easily gain access to measurement data. It would be beneficial to have the ability to channel a data stream (from an ADC perhaps) to some simple yet nice form of graphics displayed 'online'.
The Arduino IDE has such an ability but it is very restricted and not nice. Still it is useful sometimes. Would be nice to have something like this or better for MP.

Perhaps a way to display contents of a framebuffer device, which would help a lot in developing/testing framebuffer applications.

In connection with a simple hardware description file of the device (which would tell about the pin layout) a real-time display of the state of the pins could be done with different colors for the input or output setting of the pins, different levels at them and perhaps even the usage of certain pins for IO like I2C or PCI.

An option to run the REPL asynchronously, perhaps together with the [asyncio profiler](https://www.youtube.com/watch?v=qQ-t11hHWXQ) by @andrewleech.

A way to serve the IDE from the chip running MP itself.

I problably can think of a lot more. Like a display of file system usage, memory usage and fragmentation, variable inspection. Which probably are much more complicated and require additions to the MP interpreter side like those Thonny is using. Not nice, probably. So don't feel pressured by any of this.:-)

**Reply by dhalbert** [reply]
Potential feature: shadow backup version of files saved to the board. This allows direct editing of files on the board for a quick development cycle, and also easy recovery if the on-board filesystem becomes corrupted or erased.

As this is a web-based editor, it may be a bit of a user nuisance to give the permissions to access the backup location.

**Reply by vshymanskyy** [reply]
@rkompass that's some really nice ideas. Thank you!
I've been considering some of these already. These are of particular interest:
- asyncio repl
- file/folder upload and rename
- some form of data streaming/chart display


Some notes on other ideas:
> framebuffer device

no idea if this is feasible to implement on the IDE level, esp. considering that all data is pumped through a UART

> a real-time display of the state of the pins could be done with different colors

sounds like a completely separate project

> A way to serve the IDE from the chip running MP itself.

The minimal version pf ViperIDE is rather small (~2 Mb), however with all the cool features (like Ruff linter) it approaches 10 Mb.
It's not feasible to host on the device. Also, why? ViperIDE happily works (and opens) even if your PC is offline.

**Reply by rkompass** [reply]
O.k., I see.
Totally agree with your selection/preference of these 3 items.

> ViperIDE happily works (and opens) even if your PC is offline.

Didn't know that.

What else came to mind:
Perhaps offer a way to use the tools mpy-cross or even mpremote?
This might help transform .py files to .mpy (with mpy-cross) to save space and execution time+memory.

Also if one would be able to mount a directory from the PC (with mpremote), that would be awesome.

Could that be possible in principle or is it not, because you cannot run programs from the PC side?

>> framebuffer device

> no idea if this is feasible to implement on the IDE level, esp. considering that all data is pumped through a UART

perhaps it is possible using MP for Webassembly in some future time. It is not important now anyway.
Also it might be that LVGL will take over in the future.

**Comment by vshymanskyy** [comment]
The MPY disassembler can now be enabled in the latest [ViperIDE v0.4.13](https://viper-ide.org).
It works in a [very curious way](https://github.com/vshymanskyy/ViperIDE/blob/d27a5735cb0d1587105e2f845ec0bf2dda6750c3/src/python_utils.js#L153-L177)

![image](https://github.com/user-attachments/assets/6b8bd0be-768d-45b4-9a74-7a6d762b127a)
